File: internal/tools/echo.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
)

// Echo is a fake tool for M0 testing. It echoes back its input.
type Echo struct{}

func (Echo) Name() string        { return "echo" }
func (Echo) Description() string { return "Echoes back the provided input." }
func (Echo) Schema() json.RawMessage {
	return json.RawMessage(`{
		"type": "object",
		"properties": {
			"input": {
				"type": "string",
				"description": "The string to echo back."
			}
		},
		"required": ["input"]
	}`)
}

func (Echo) Execute(_ context.Context, input map[string]any) (Result, error) {
	v, ok := input["input"].(string)
	if !ok {
		return Result{Error: "missing or non-string input"}, nil
	}
	return Result{Output: fmt.Sprintf("echo: %s", v)}, nil
}

File: crates/mew-tools/src/tools/echo.rs
use crate::{Sensitivity, Tool, ToolCtx, ToolError, ToolOutput};
use async_trait::async_trait;
use serde_json::Value;

pub struct Echo;

#[async_trait]
impl Tool for Echo {
    fn name(&self) -> &str {
        "echo"
    }

    fn description(&self) -> &str {
        "Echoes back the provided input."
    }

    fn schema(&self) -> &Value {
        static SCHEMA: std::sync::OnceLock<Value> = std::sync::OnceLock::new();
        SCHEMA.get_or_init(|| {
            serde_json::json!({
                "type": "object",
                "properties": {
                    "input": {
                        "type": "string",
                        "description": "The string to echo back."
                    }
                },
                "required": ["input"]
            })
        })
    }

    fn sensitivity(&self) -> Sensitivity {
        Sensitivity::ReadOnly
    }

    async fn execute(&self, _ctx: ToolCtx, input: Value) -> Result<ToolOutput, ToolError> {
        let s = input
            .get("input")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::InvalidInput("missing or non-string input".into()))?;

        Ok(ToolOutput {
            output: format!("echo: {}", s),
            error: String::new(),
            diff: None,
        })
    }
}
